velox/expression/tests/ExprOptimizerTest.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <gtest/gtest.h>
+
+#include "velox/expression/ExprConstants.h"
+#include "velox/expression/ExprOptimizer.h"
+#include "velox/expression/ExprRewriteRegistry.h"
+#include "velox/functions/prestosql/registration/RegistrationFunctions.h"
+#include "velox/parse/Expressions.h"
+#include "velox/parse/ExpressionsParser.h"
+#include "velox/parse/TypeResolver.h"
+#include "velox/vector/tests/utils/VectorTestBase.h"
+
+namespace facebook::velox::expression::test {
+
+class ExprOptimizerTest : public testing::Test,
+                          public velox::test::VectorTestBase {
+ protected:
+  static void SetUpTestCase() {
+    memory::MemoryManager::testingSetInstance(memory::MemoryManager::Options{});
+  }
+
+  void SetUp() override {
+    functions::prestosql::registerAllScalarFunctions("");
+    parse::registerTypeResolver();
+  }
+
+  void TearDown() override {
+    expression::ExprRewriteRegistry::instance().clear();
+  }
+
+  core::TypedExprPtr makeTypedExpr(
+      const std::string& expression,
+      const RowTypePtr& type) {
+    auto untyped = parse::parseExpr(expression, {});
+    return core::Expressions::inferTypes(untyped, type, execCtx_->pool());
+  }
+
+  core::TypedExprPtr optimize(
+      const std::string& expression,
+      const RowTypePtr& type = ROW({}),
+      bool replaceEvalErrorWithFailExpr = false) {
+    const auto expr = makeTypedExpr(expression, type);
+    return expression::optimize(
+        expr, queryCtx_.get(), pool(), replaceEvalErrorWithFailExpr);
+  }
+
+  void testExpression(
+      const std::string& input,
+      const std::string& expected,
+      const RowTypePtr& inputType = ROW({}),
+      const RowTypePtr& expectedType = ROW({})) {
+    auto optimized = optimize(input, inputType);
+    auto optimizedExpected = optimize(expected, expectedType);
+    // The constant value in ConstantTypedExpr can either be in valueVector_ or
+    // in the variant value_. String comparison is used to compare the optimized
+    // expressions, since ITypedExpr comparison with equality operator will fail
+    // for `ConstantTypedExpr`s if the underlying constant values do not have
+    // the same internal representation.
+    ASSERT_EQ(optimized->toString(), optimizedExpected->toString());
+  }
+
+  void testFailExpression(
```

Comment:
This method doesn't seem to be testing anything, rather it checks if 'expr' is 'fail(expectedError)'. Is this so? It might be simpler to just compare expr->toString() with "fail(expectedError)" or similar.

Reply:
Reverted, thanks.